backend/src/types.ts

```diff
@@ -666,3 +659,62 @@ type GroupCustomObjectItemMetadata = {
 export type RecursivePartial<T> = {
   [P in keyof T]?: RecursivePartial<T[P]>;
 };
+
+export type EnvironmentVariable = EitherNotBoth<
+  { value: string | number },
+  { valueFrom: Record<string, unknown> }
+> & {
+  name: string;
+};
+
+export type ResourceGetter<T extends K8sResourceCommon> = (
+  fastify: KubeFastifyInstance,
+  namespace: string,
+  name: string,
+) => Promise<T>;
+
+export type ResourceConstructor<T extends K8sResourceCommon> = (
+  namespace: string,
+  name: string,
+  resource: Record<string, string>,
+) => T;
+
+export type ResourceCreator<T extends K8sResourceCommon> = (
+  fastify: KubeFastifyInstance,
+  resource: T,
+) => Promise<T>;
+
+export type ResourceUpdater<T extends K8sResourceCommon> = (
+  fastify: KubeFastifyInstance,
+  resource: T,
+) => Promise<T>;
+
+export type EnvVarResource = V1Secret | V1ConfigMap;
+
+export type EnvVarReducedTypeKeyValues = {
+  configMap: Record<string, string>;
+  secrets: Record<string, string>;
+};
+
+export enum EnvVarResourceType {
+  Secret = 'Secret',
+  ConfigMap = 'ConfigMap',
+}
+
+export enum NotebookState {
+  Started = 'started',
+  Stopped = 'stopped',
+}
+
+export type NotebookData = {
+  notebookSizeName: string;
+  imageName: string;
+  imageTagName: string;
+  url: string;
+  gpus: number;
+  envVars: EnvVarReducedTypeKeyValues;
+  state?: NotebookState;
```

Comment:
```suggestion
  state: NotebookState;
```
Couldn't find a situation where this being optional was helpful/used. Was this just a transition in the code as you developed?

Reply:
oh, i was doing the refactor with all the improvements and I ended up adding that same thing. I made it optional when we were discussing the way to implement the stop and start, but it ended up being mandatory. tx!